src/extensions/default/AutoUpdate/unittests.js

```diff
@@ -0,0 +1,491 @@
+                btn;
+
+            // Test if the update bar button has been displayed.
+            expect(doc.find(UPDATE_BAR).length).toBe(1);
+            btn = doc.find(btnId);
+            expect(btn.length).toBe(1);
+
+            if (btn && click) {
+                // We will wait for 3 seconds so that the Update Bar is visible to the tester.
+                waitForMilliSeconds(3000);
+                btn.click();
+                setTimeout(function () {
+                    // We will wait for 1 second for the click action to take effect
+                    expect(doc.find(UPDATE_BAR).length).toBe(0);
+                    updateBarDismissed = true;
+                }, 5000);
+            }
+        }
+
+        /**
+         * Check the presence of Update Bar String on Brackets Window
+         * @param {String} title - Title String Which will be matched with Update Bar heading.
+         * @param {String} description - description String Which will be matched with Update Bar description.
+         */
+        function checkUpdateBarString(title, titleDescription) {
+            var doc = $(testWindow.document),
+                updBar = doc.find(UPDATE_BAR),
+                heading = updBar.find('#heading'),
+                description = updBar.find('#description');
+
+            // Test if the update bar has been displayed.
+            expect(updBar.length).toBe(1);
+            if (title) {
+                expect(heading.text()).toBe(title);
+            }
+            if (titleDescription) {
+                expect(description.text()).toBe(titleDescription);
+            }
+        }
+
+        /**
+         * Check the presence of Update Bar Prompt on Brackets Window
+         */
+        function checkUpdateBar() {
+            var doc = $(testWindow.document),
+                updBar = doc.find(UPDATE_BAR),
+                updBarPresent = false;
+
+            if (updBar && updBar.length > 0) {
+                updBarPresent =  true;
+            }
+            return updBarPresent;
+        }
+
+        it("AutoUpdate Exiension should be loaded Successfuly", function () {
+            waitForMilliSeconds(5000);
+            runs(function () {
+                expect(AutoUpdate).not.toBeNull();
+                if (AutoUpdate) {
+                    expect(AutoUpdate._updateProcessHandler).not.toBeNull();
+                    expect(AutoUpdate.checkUpdateStatus).not.toBeNull();
+                    expect(AutoUpdate.setUpdateStateInJSON).not.toBeNull();
+                    expect(AutoUpdate.getPlatformInfo).not.toBeNull();
+                }
+            });
+        });
+
+
+        it("should attempt download, fail, retry and eventually prompt for failure", downloadFailure);
+
+        /**
+         * Tests the download failure scenario for Auto Update
+         * and validate download failure message in Update Bar prompt.
+         */
+        function downloadFailure() {
```

Comment:
Is the intension to simulate a failure scenario or just check if their are any failures? The former one makes more sense and can be simulated by giving a wrong URL.

Reply:
intension is to simulate failure scenario by giving a wrong URL for download failure and passing invalid checksum value for validation failure.